# Webscrapping

## Problem Statement

In 2020, Covid-19 has greatly impacted many countries which resulted in mulitple countries going into lockdown where physical stores are force to shut their doors and many businesses moved online. As such, ecommerce sales skyrocketed and account for [101% of all growth in retail in US](https://www.digitalcommerce360.com/article/us-ecommerce-sales/)<br><br>
As many small businesses does not have the necessary resources nor expertise to setup and manage their own website, online platform such as Etsy and Shopify are able to provide an effective platform to help these businesses quickly and easily transit into ecommerce. Despite both website seemingly provide the same service, it is in fact 2 different kindly of ecommerce platform. Etsy is an online market place where most transaction will occur there, while Shopify is a webstore platform that help to build a website for an online business.
<br><br>
The objective of this project is to create a classification model to help effectively categorize the subreddit post for Etsy and Shopify and discover what are the main differences to help small businesses select the most effective platform for themselves. It would also give an understand the main focus of our client so as to effectively channel our market strategies to increase Etsy market share.

In [1]:
import pandas as pd
import requests
import time
import random
import string

## Scrapping Shopify Subreddit

In [53]:
#To randomly create an agent for each iteration
def agent():
    return"".join([(random.choice(string.ascii_uppercase + string.digits)) for i in range(random.randint(10,15))])

In [3]:
url_shop= 'http://www.reddit.com/r/Shopify.json'

In [4]:
post = []
after= None

for a in range(40):
    if after == None:
        current_url=url_shop
    else:
        current_url= url_shop + '?after=' + after
    print(current_url)
    res =requests.get(current_url,headers={'User-agent': agent()})
    
    if res.status_code != 200:
        print('Status error',res.status_code)
        break
    
    current_dict = res.json()
    current_posts = [p['data'] for p in current_dict['data']['children']]
    post.extend(current_posts)
    after =current_dict['data']['after']

    
    sleep_duration= random.randint(2,10)
    print(sleep_duration)
    time.sleep(sleep_duration)

http://www.reddit.com/r/Shopify.json
5
http://www.reddit.com/r/Shopify.json?after=t3_m24n6p
9
http://www.reddit.com/r/Shopify.json?after=t3_m1msdo
2
http://www.reddit.com/r/Shopify.json?after=t3_m06gn8
5
http://www.reddit.com/r/Shopify.json?after=t3_m0396v
9
http://www.reddit.com/r/Shopify.json?after=t3_m00rap
3
http://www.reddit.com/r/Shopify.json?after=t3_lzajx3
8
http://www.reddit.com/r/Shopify.json?after=t3_lyjirv
7
http://www.reddit.com/r/Shopify.json?after=t3_ly38hd
10
http://www.reddit.com/r/Shopify.json?after=t3_lx8umr
6
http://www.reddit.com/r/Shopify.json?after=t3_lwmvko
6
http://www.reddit.com/r/Shopify.json?after=t3_lwd5vh
3
http://www.reddit.com/r/Shopify.json?after=t3_lvlodl
7
http://www.reddit.com/r/Shopify.json?after=t3_luwk48
10
http://www.reddit.com/r/Shopify.json?after=t3_lu16y9
9
http://www.reddit.com/r/Shopify.json?after=t3_lsxv9e
7
http://www.reddit.com/r/Shopify.json?after=t3_lski4d
5
http://www.reddit.com/r/Shopify.json?after=t3_lrwzi1
6
http://www.reddit.com/r/

From the printed URL, it can be seen that it was loop back to the original URL and hence there will be duplicated post that was scrapped

In [5]:
shopify=pd.DataFrame(post)

In [6]:
shopify.drop_duplicates(subset=['selftext'],inplace=True)

In [15]:
shopify

approved_at_utc subreddit  \
0              None   shopify   
1              None   shopify   
2              None   shopify   
3              None   shopify   
4              None   shopify   
..              ...       ...   
763            None   shopify   
764            None   shopify   
765            None   shopify   
766            None   shopify   
767            None   shopify   

                                              selftext author_fullname  saved  \
0    **As the sub has changed over the last few wee...        t2_kie3p  False   
1    **Post store, service and app promotion messag...        t2_wa5w5  False   
2    I posted this before on another sub, but I’d a...        t2_bfsfs  False   
3    Hi all,\nWhat are you all using for landing pa...     t2_97ailqyo  False   
4    There are hundreds of timer apps on Shopify's ...     t2_5e6typm4  False   
..                                                 ...             ...    ...   
763  Hi guys!\n\nI am launching a service, where pe...     t2_2xhy6fsq  False   
764  Hi guys, I own a gift hamper store and I’m loo...       t2_13c4cm  False   
765  HI! Welcome! We are a small business for those...      t2_9ujl9xs  False   
766  Hey guys,\n\nI was just wondering is it a good...     t2_5qxtvisl  False   
767  I was wondering if anyone has experience with ...     t2_39k3rh1o  False   

    mod_reason_title  gilded  clicked  \
0               None       0    False   
1               None       0    False   
2               None       0    False   
3               None       0    False   
4               None       0    False   
..               ...     ...      ...   
763             None       0    False   
764             None       0    False   
765             None       0    False   
766             None       0    False   
767             None       0    False   

                                                 title  \
0                                    r/Shopify Updates   
1          Promote your Shopify business, weekly post.   
2    The best 5 marketing tools that worked for our...   
3                                        Landing pages   
4    Any legit timer countdown apps where you set t...   
..                                                 ...   
763              Is Shopify suitable for this usecase?   
764    Need some help with inventory and stock control   
765                                PROSPECT APOTHECARY   
766                                   Selling posters?   
767                               Valentines day slow?   

                                   link_flair_richtext  ... num_crossposts  \
0                 [{'e': 'text', 't': 'Announcement'}]  ...              0   
1               [{'e': 'text', 't': 'Self Promotion'}]  ...              0   
2                    [{'e': 'text', 't': 'Marketing'}]  ...              0   
3                    [{'e': 'text', 't': 'Marketing'}]  ...              0   
4                         [{'e': 'text', 't': 'Apps'}]  ...              0   
..                                                 ...  ...            ...   
763  [{'e': 'text', 't': 'Shopify General Discussio...  ...              0   
764                       [{'e': 'text', 't': 'Apps'}]  ...              0   
765                   [{'e': 'text', 't': 'Products'}]  ...              0   
766                   [{'e': 'text', 't': 'Products'}]  ...              0   
767  [{'e': 'text', 't': 'Shopify General Discussio...  ...              0   

     media  is_video post_hint  \
0     None     False       NaN   
1     None     False       NaN   
2     None     False      self   
3     None     False       NaN   
4     None     False       NaN   
..     ...       ...       ...   
763   None     False       NaN   
764   None     False       NaN   
765   None     False       NaN   
766   None     False       NaN   
767   None     False       NaN   

                                               preview  poll_data  \
0                            

After dropping all the duplicated post. We have 765 rows remaining which would be sufficient for our modeling

In [17]:
# Some of the title might be able to give us a good insight and hence we will combine the selftext with the title
shopify['text']=shopify['selftext']+shopify['title']

In [19]:
shopify.columns

Index(['approved_at_utc', 'subreddit', 'selftext', 'author_fullname', 'saved',
       'mod_reason_title', 'gilded', 'clicked', 'title', 'link_flair_richtext',
       ...
       'media', 'is_video', 'post_hint', 'preview', 'poll_data',
       'crosspost_parent_list', 'url_overridden_by_dest', 'crosspost_parent',
       'author_cakeday', 'text'],
      dtype='object', length=113)

In [25]:
shopify.drop(columns=[column for column in shopify.columns if (column !='text')&(column !='subreddit')],inplace=True)

In [50]:
shopify.to_csv('./data/shopify.csv')

## Scrapping Etsy Subreddit

In [8]:
url_etsy= 'http://www.reddit.com/r/Etsy.json'

In [9]:
post = []
after= None

for a in range(40):
    if after == None:
        current_url=url_etsy
    else:
        current_url= url_etsy + '?after=' + after
    print(current_url)
    res =requests.get(current_url,headers={'User-agent': agent()})
    
    if res.status_code != 200:
        print('Status error',res.status_code)
        break
    
    current_dict = res.json()
    current_posts = [p['data'] for p in current_dict['data']['children']]
    post.extend(current_posts)
    after =current_dict['data']['after']

    
    sleep_duration= random.randint(2,10)
    print(sleep_duration)
    time.sleep(sleep_duration)

http://www.reddit.com/r/Etsy.json
5
http://www.reddit.com/r/Etsy.json?after=t3_m24f9p
9
http://www.reddit.com/r/Etsy.json?after=t3_m1ar4g
7
http://www.reddit.com/r/Etsy.json?after=t3_m0p3ac
8
http://www.reddit.com/r/Etsy.json?after=t3_lzr7w9
10
http://www.reddit.com/r/Etsy.json?after=t3_lzdhon
4
http://www.reddit.com/r/Etsy.json?after=t3_lyus7x
8
http://www.reddit.com/r/Etsy.json?after=t3_lyd96z
3
http://www.reddit.com/r/Etsy.json?after=t3_lx4jr7
4
http://www.reddit.com/r/Etsy.json?after=t3_lx567j
4
http://www.reddit.com/r/Etsy.json?after=t3_lwfur2
5
http://www.reddit.com/r/Etsy.json?after=t3_lva6r5
6
http://www.reddit.com/r/Etsy.json?after=t3_lutiai
9
http://www.reddit.com/r/Etsy.json?after=t3_lub6hm
7
http://www.reddit.com/r/Etsy.json?after=t3_ltqbns
2
http://www.reddit.com/r/Etsy.json?after=t3_lt0fi6
5
http://www.reddit.com/r/Etsy.json?after=t3_ls3lag
2
http://www.reddit.com/r/Etsy.json?after=t3_lrp05c
7
http://www.reddit.com/r/Etsy.json?after=t3_lrb4u0
3
http://www.reddit.com/r/Ets

In [12]:
etsy=pd.DataFrame(post)

In [13]:
etsy.drop_duplicates(subset=['selftext'],inplace=True)

In [14]:
etsy

approved_at_utc subreddit  \
0              None      Etsy   
1              None      Etsy   
2              None      Etsy   
3              None      Etsy   
4              None      Etsy   
..              ...       ...   
993            None      Etsy   
994            None      Etsy   
995            None      Etsy   
996            None      Etsy   
997            None      Etsy   

                                              selftext author_fullname  saved  \
0    Hi! Welcome to the monthly critique request th...        t2_kn6jk  False   
1                                                             t2_kn6jk  False   
2    Hi guys! So I sell mostly press on nails, lip ...     t2_9z3i08w0  False   
3    Shipped item on time. Buyer upset it is late. ...        t2_8ua9r  False   
4    Hello, I placed an order 1/15/21 that was supp...       t2_10od41  False   
..                                                 ...             ...    ...   
993  I bought an item from a etsy seller in the ver...     t2_76z5aonw  False   
994  Hi I was wondering if anyone could give me som...     t2_2dh8no9s  False   
995  On my homepage, I keep seeing explicit art pri...     t2_3pvdpwq8  False   
996  Way back in late September, I bought something...     t2_9szwc5mw  False   
997  Recently I noticed how the amount of attention...     t2_9f3zo7wp  False   

    mod_reason_title  gilded  clicked  \
0               None       0    False   
1               None       0    False   
2               None       0    False   
3               None       0    False   
4               None       0    False   
..               ...     ...      ...   
993             None       0    False   
994             None       0    False   
995             None       0    False   
996             None       0    False   
997             None       0    False   

                                                 title  \
0    Critique, Help, and Feedback Requests - March ...   
1         Share Your Stuff - March 8th thru 14th, 2021   
2    Do customers like personal touches to their pa...   
3           Etsy sided with seller's 1-star review....   
4    When do I open a case for a refund? When is en...   
..                                                 ...   
993              What is trust and safety review like?   
994  Order dispatched (Royal mail) but both product...   
995                    Is there a way to block a shop?   
996                       Need refund from closed shop   
997                           I badly need comfort lol   

                       link_flair_richtext  ... num_crossposts  media  \
0                                       []  ...              0   None   
1                                       []  ...              0   None   
2    [{'e': 'text', 't': 'Advice Needed'}]  ...              0   None   
3                                       []  ...              0   None   
4    [{'e': 'text', 't': 'Advice Needed'}]  ...              0   None   
..                                     ...  ...            ...    ...   
993                                     []  ...              0   None   
994           [{'e': 'text', 't': 'Help'}]  ...              0   None   
995                                     []  ...              0   None   
996                                     []  ...              0   None   
997  [{'e': 'text', 't': 'Advice Needed'}]  ...              0   None   

     is_video                link_flair_template_id  post_hint preview  \
0       False                                   NaN        NaN     NaN   
1       False                                   NaN        NaN     NaN   
2       False  3b325b76-6c9f-11ea-ab90-0e9af179a2e7        NaN     NaN   
3       False                                   NaN        NaN     NaN   
4       False  3b325b76-6c9f-11ea-ab90-0e9af179a2e7        NaN     NaN   
..        ...                                   ...        ...     ...   
993     False                                   N

In [26]:
etsy['text']=etsy['selftext']+etsy['title']

In [29]:
etsy.drop(columns=[column for column in etsy.columns if (column !='text')&(column !='subreddit')],inplace=True)

In [30]:
etsy

subreddit                                               text
0        Etsy  Hi! Welcome to the monthly critique request th...
1        Etsy       Share Your Stuff - March 8th thru 14th, 2021
2        Etsy  Hi guys! So I sell mostly press on nails, lip ...
3        Etsy  Shipped item on time. Buyer upset it is late. ...
4        Etsy  Hello, I placed an order 1/15/21 that was supp...
..        ...                                                ...
993      Etsy  I bought an item from a etsy seller in the ver...
994      Etsy  Hi I was wondering if anyone could give me som...
995      Etsy  On my homepage, I keep seeing explicit art pri...
996      Etsy  Way back in late September, I bought something...
997      Etsy  Recently I noticed how the amount of attention...

[993 rows x 2 columns]

In [52]:
etsy.to_csv('./data/etsy.csv')

## Combining Dataset

In [43]:
combine = pd.concat([etsy,shopify])

In [44]:
combine

subreddit                                               text
0        Etsy  Hi! Welcome to the monthly critique request th...
1        Etsy       Share Your Stuff - March 8th thru 14th, 2021
2        Etsy  Hi guys! So I sell mostly press on nails, lip ...
3        Etsy  Shipped item on time. Buyer upset it is late. ...
4        Etsy  Hello, I placed an order 1/15/21 that was supp...
..        ...                                                ...
763   shopify  Hi guys!\n\nI am launching a service, where pe...
764   shopify  Hi guys, I own a gift hamper store and I’m loo...
765   shopify  HI! Welcome! We are a small business for those...
766   shopify  Hey guys,\n\nI was just wondering is it a good...
767   shopify  I was wondering if anyone has experience with ...

[1758 rows x 2 columns]

In [45]:
combine.to_csv('./data/combine.csv',index=False)